_posts/2023-08-14-clear-specification-and-implementation.markdown

```diff
@@ -0,0 +1,610 @@
+  // then popping yields the pushed value.
+  lemma PushPopVal<T>(stk: StackModel<T>, val : T)
+    ensures PopModel(PushModel(stk, val)).0 == val {}
+
+  // The second core property of stacks, for use by clients. Pushing and
+  // then popping yields the original stack.
+  lemma PushPopStack<T>(stk: StackModel<T>, val : T)
+    ensures PopModel(PushModel(stk, val)).1 == stk {}
+}
+{% endhighlight %}
+
+Several things are important about this implementation. First, it
+provides a number of functions with bodies that describe one way of
+constructing a stack out of built-in primitives. Second, it hides
+the bodies of those functions from most clients, except
+implementation modules that need to prove equivalence with them.
+And, third, it describes the _relationship_ between those operations
+in a couple of lemmas.
+
+One further note: you'll see that some of the operations on a stack
+have pre- and post-conditions, specified on the functions themselves
+rather than in lemmas. These are exactly the properties required to
+ensure that each operation is well-defined without specifying
+anything else about behavior. Besides well-definedness, all
+functional properties are specified in lemmas.
+
+Now let's look at how a common implementation of this data type
+might be structured. This approach uses a consecutive array to store
+elements, which we look at first because it's somewhat simpler than
+the linked-list implementation we'll consider next.
+
+Before digging into the implementation, there's one subtlety we need
+to discuss. The array-based approach sometimes requires allocating
+more space than is currently used, making it necessary to have a
+value available to fill unused entries. Many Dafny types have
+default values that can be used for this purpose, but not all types
+do. It is possible to constrain type parameters to those that do
+have default values by, for instance, using the declaration `class
+Stack<T(0)>`, to insist that `T` have type characteristic `0`, the
+notation for having a default value.
+
+However, doing this would make the array-based stack have different
+constraints on its type parameter than other stack implementations,
+which isn't ideal. An alternative is to make use of the fact that Dafny
+can define datatypes with ghost alternatives, and that datatypes with
+only one non-ghost alternative with only one argument have no runtime
+overhead. Therefore, the array-based stack implementation uses an
+(internal) datatype called `GhostOption` that makes it possible for
+`Stack` to be parameterized by an arbitrary type because the
+`GhostOption` type always has a default value.
+
+{% highlight dafny %}
+module ArrayStackImplementation {
+  // Get access to the implementation details.
+  import opened StackSpecification`Private
+
+  // This allows `elts` below to include default values in unused
+  // positions.
+  datatype GhostOption<T> = ghost None | Some(value: T)
```

Comment:
Wow that's clever! But practically speaking, we still store null? Can you explain?

Reply:
I believe the compiler can leave the storage uninitialized, because Dafny can prove that it's never accessed (because that would require using a ghost constructor at runtime).